File: ios/SpentyAI/SpentyAI/Features/Records/RecordsViewModel.swift
import Foundation
import Observation
import SwiftUI

enum RecordsTab: String, CaseIterable {
    case emails = "Emails"
    case receipts = "Receipts"
}

@Observable
final class RecordsViewModel {

    // MARK: - Data

    var records: [Record] = []
    var receipts: [Receipt] = []
    var recordsTotal: Int = 0
    var receiptsTotal: Int = 0

    // MARK: - Tab

    var activeTab: RecordsTab = .emails

    // MARK: - Filters (Emails)

    var searchQuery: String = ""
    var dateFrom: Date? = nil
    var dateTo: Date? = nil
    var amountMin: String = ""
    var amountMax: String = ""

    // MARK: - UI State

    var isLoading: Bool = false
    var isLoadingMore: Bool = false
    var page: Int = 1
    var hasMore: Bool = true

    var receiptsPage: Int = 1
    var receiptsHasMore: Bool = true
    var isLoadingReceipts: Bool = false
    var isLoadingMoreReceipts: Bool = false

    var selectedRecord: Record? = nil
    var errorMessage: String? = nil

    var isDownloadingZip: Bool = false
    var isDownloadingEml: Bool = false

    var shareItem: ShareItem? = nil

    // MARK: - Private

    private let repository = RecordsRepository.shared
    private let pageSize = 30

    private static let queryDateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    // MARK: - Record Operations

    @MainActor
    func loadRecords() async {
        isLoading = true
        errorMessage = nil
        page = 1
        do {
            let response = try await fetchRecordsPage(page: 1)
            records = response.records
            recordsTotal = response.total
            hasMore = records.count < recordsTotal
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    @MainActor
    func loadMoreRecords() async {
        guard !isLoadingMore, hasMore else { return }
        isLoadingMore = true
        page += 1
        do {
            let response = try await fetchRecordsPage(page: page)
            records.append(contentsOf: response.records)
            hasMore = records.count < recordsTotal
        } catch {
            page -= 1
            errorMessage = error.localizedDescription
        }
        isLoadingMore = false
    }

    @MainActor
    func refreshRecords() async {
        page = 1
        errorMessage = nil
        do {
            let response = try await fetchRecordsPage(page: 1)
            records = response.records
            recordsTotal = response.total
            hasMore = records.count < recordsTotal
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func searchRecords() async {
        guard !searchQuery.isEmpty else {
            await refreshRecords()
            return
        }
        isLoading = true
        errorMessage = nil
        do {
            let response = try await repository.searchRecords(query: searchQuery)
            records = response.records
            recordsTotal = response.total
            hasMore = false
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    @MainActor
    func deleteRecord(id: String) async {
        do {
            _ = try await repository.deleteRecord(id: id)
            records.removeAll { $0.id == id }
            recordsTotal -= 1
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func downloadEml(id: String, subject: String?) async {
        isDownloadingEml = true
        do {
            let data = try await repository.downloadEml(id: id)
            let filename = (subject ?? "email").replacingOccurrences(of: "/", with: "-") + ".eml"
            let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent(filename)
            try data.write(to: tempURL)
            shareItem = ShareItem(url: tempURL)
        } catch {
            errorMessage = error.localizedDescription
        }
        isDownloadingEml = false
    }

    @MainActor
    func downloadAttachment(recordId: String, index: Int, filename: String?) async {
        do {
            let data = try await repository.downloadAttachment(id: recordId, index: index)
            let name = filename ?? "attachment_\(index)"
            let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent(name)
            try data.write(to: tempURL)
            shareItem = ShareItem(url: tempURL)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func downloadZip() async {
        isDownloadingZip = true
        do {
            let data = try await repository.downloadZip()
            let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent("records_export.zip")
            try data.write(to: tempURL)
            shareItem = ShareItem(url: tempURL)
        } catch {
            errorMessage = error.localizedDescription
        }
        isDownloadingZip = false
    }

    // MARK: - Receipt Operations

    @MainActor
    func loadReceipts() async {
        isLoadingReceipts = true
        errorMessage = nil
        receiptsPage = 1
        do {
            let response = try await repository.fetchReceipts(page: 1, limit: pageSize)
            receipts = response.receipts
            receiptsTotal = response.total
            receiptsHasMore = receipts.count < receiptsTotal
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoadingReceipts = false
    }

    @MainActor
    func loadMoreReceipts() async {
        guard !isLoadingMoreReceipts, receiptsHasMore else { return }
        isLoadingMoreReceipts = true
        receiptsPage += 1
        do {
            let response = try await repository.fetchReceipts(page: receiptsPage, limit: pageSize)
            receipts.append(contentsOf: response.receipts)
            receiptsHasMore = receipts.count < receiptsTotal
        } catch {
            receiptsPage -= 1
            errorMessage = error.localizedDescription
        }
        isLoadingMoreReceipts = false
    }

    @MainActor
    func refreshReceipts() async {
        receiptsPage = 1
        errorMessage = nil
        do {
            let response = try await repository.fetchReceipts(page: 1, limit: pageSize)
            receipts = response.receipts
            receiptsTotal = response.total
            receiptsHasMore = receipts.count < receiptsTotal
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func deleteReceipt(id: String) async {
        do {
            _ = try await repository.deleteReceipt(id: id)
            receipts.removeAll { $0.id == id }
            receiptsTotal -= 1
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func downloadReceipt(id: String, filename: String?) async {
        do {
            let data = try await repository.downloadReceipt(id: id)
            let name = filename ?? "receipt_\(id)"
            let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent(name)
            try data.write(to: tempURL)
            shareItem = ShareItem(url: tempURL)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    func uploadReceipt(imageData: Data, filename: String, mimeType: String) async -> String? {
        do {
            let response = try await repository.uploadReceipt(
                imageData: imageData,
                filename: filename,
                mimeType: mimeType
            )
            await loadReceipts()
            return response.id
        } catch {
            errorMessage = error.localizedDescription
            return nil
        }
    }

    @MainActor
    func parseReceipt(id: String) async -> Receipt? {
        do {
            let parsed = try await repository.parseReceipt(id: id)
            if let idx = receipts.firstIndex(where: { $0.id == id }) {
                receipts[idx] = parsed
            }
            return parsed
        } catch {
            errorMessage = error.localizedDescription
            return nil
        }
    }

    @MainActor
    func linkReceiptToTransaction(receiptId: String, transactionId: String) async {
        do {
            let updated = try await repository.linkReceipt(id: receiptId, transactionId: transactionId)
            if let idx = receipts.firstIndex(where: { $0.id == receiptId }) {
                receipts[idx] = updated
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Helpers

    private func fetchRecordsPage(page: Int) async throws -> RecordListResponse {
        let dateFromStr = dateFrom.map { Self.queryDateFormatter.string(from: $0) }
        let dateToStr = dateTo.map { Self.queryDateFormatter.string(from: $0) }
        let minAmt = Double(amountMin)
        let maxAmt = Double(amountMax)

        return try await repository.fetchRecords(
            page: page,
            limit: pageSize,
            dateFrom: dateFromStr,
            dateTo: dateToStr,
            amountMin: minAmt,
            amountMax: maxAmt
        )
    }
}

// MARK: - Share Item

struct ShareItem: Identifiable {
    let id = UUID()
    let url: URL
}

File: android-native/app/src/main/java/com/spentyai/app/features/records/RecordsViewModel.kt
package com.spentyai.app.features.records

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spentyai.app.core.models.*
import com.spentyai.app.core.network.ApiResult
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*

enum class RecordsTab(val label: String) {
    EMAILS("Emails"),
    RECEIPTS("Receipts")
}

class RecordsViewModel(private val repository: RecordsRepository) : ViewModel() {

    // Data
    var records by mutableStateOf<List<Record>>(emptyList()); private set
    var receipts by mutableStateOf<List<Receipt>>(emptyList()); private set
    var recordsTotal by mutableStateOf(0); private set
    var receiptsTotal by mutableStateOf(0); private set

    // Tab
    var activeTab by mutableStateOf(RecordsTab.EMAILS)

    // Filters (Emails)
    var searchQuery by mutableStateOf("")
    var dateFrom by mutableStateOf<Date?>(null)
    var dateTo by mutableStateOf<Date?>(null)
    var amountMin by mutableStateOf("")
    var amountMax by mutableStateOf("")

    // UI State
    var isLoading by mutableStateOf(false); private set
    var isLoadingMore by mutableStateOf(false); private set
    var hasMore by mutableStateOf(true); private set

    var isLoadingReceipts by mutableStateOf(false); private set
    var isLoadingMoreReceipts by mutableStateOf(false); private set
    var receiptsHasMore by mutableStateOf(true); private set

    var errorMessage by mutableStateOf<String?>(null)
    var successMessage by mutableStateOf<String?>(null)

    var isDownloadingZip by mutableStateOf(false); private set
    var isDownloadingEml by mutableStateOf(false); private set

    // Record Preview State
    var activePreview by mutableStateOf<RecordPreviewResponse?>(null); private set
    var isLoadingPreview by mutableStateOf(false); private set
    var previewError by mutableStateOf<String?>(null); private set

    // Receipt Upload State
    var isUploading by mutableStateOf(false); private set
    var isParsing by mutableStateOf(false); private set
    var uploadedReceiptId by mutableStateOf<String?>(null)
    var parsedReceipt by mutableStateOf<Receipt?>(null)
    var linkedTransactionId by mutableStateOf("")

    private val pageSize = 30
    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    val hasDateFilter: Boolean get() = dateFrom != null || dateTo != null
    val hasAmountFilter: Boolean get() = amountMin.isNotEmpty() || amountMax.isNotEmpty()

    // Record Operations
    fun loadRecords() {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null

            when (val r = fetchRecordsPage(0)) {
                is ApiResult.Success -> {
                    records = r.data.records
                    recordsTotal = r.data.total
                    hasMore = records.size < recordsTotal
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            isLoading = false
        }
    }

    fun loadMoreRecords() {
        if (isLoadingMore || !hasMore) return
        viewModelScope.launch {
            isLoadingMore = true

            when (val r = fetchRecordsPage(records.size)) {
                is ApiResult.Success -> {
                    records = records + r.data.records
                    hasMore = records.size < recordsTotal
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            isLoadingMore = false
        }
    }

    fun refreshRecords() {
        viewModelScope.launch {
            errorMessage = null

            when (val r = fetchRecordsPage(0)) {
                is ApiResult.Success -> {
                    records = r.data.records
                    recordsTotal = r.data.total
                    hasMore = records.size < recordsTotal
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }
        }
    }

    fun searchRecords() {
        if (searchQuery.isEmpty()) {
            refreshRecords()
            return
        }
        viewModelScope.launch {
            isLoading = true
            errorMessage = null

            when (val r = repository.searchRecords(searchQuery)) {
                is ApiResult.Success -> {
                    records = r.data.items
                    recordsTotal = r.data.total
                    hasMore = false
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            isLoading = false
        }
    }

    fun deleteRecord(id: String) {
        viewModelScope.launch {
            when (repository.deleteRecord(id)) {
                is ApiResult.Success -> {
                    records = records.filter { it.id != id }
                    recordsTotal -= 1
                }
                is ApiResult.Failure -> errorMessage = "Failed to delete record"
            }
        }
    }

    fun downloadZip() {
        viewModelScope.launch {
            isDownloadingZip = true
            val archiveIds = records.map { it.id }
            when (repository.downloadZip(archiveIds)) {
                is ApiResult.Success -> successMessage = "ZIP download started"
                is ApiResult.Failure -> errorMessage = "Failed to download ZIP"
            }
            isDownloadingZip = false
        }
    }

    fun downloadEml(id: String) {
        viewModelScope.launch {
            isDownloadingEml = true
            when (repository.downloadEml(id)) {
                is ApiResult.Success -> successMessage = "EML download started"
                is ApiResult.Failure -> errorMessage = "Failed to download EML"
            }
            isDownloadingEml = false
        }
    }

    fun downloadAttachment(recordId: String, index: Int) {
        viewModelScope.launch {
            when (repository.downloadAttachment(recordId, index)) {
                is ApiResult.Success -> successMessage = "Attachment download started"
                is ApiResult.Failure -> errorMessage = "Failed to download attachment"
            }
        }
    }

    // Record Preview
    fun loadRecordPreview(id: String) {
        viewModelScope.launch {
            isLoadingPreview = true
            previewError = null
            activePreview = null

            when (val r = repository.fetchRecordPreview(id)) {
                is ApiResult.Success -> activePreview = r.data
                is ApiResult.Failure -> previewError = r.error.message
            }

            isLoadingPreview = false
        }
    }

    // Receipt Operations
    fun loadReceipts() {
        viewModelScope.launch {
            isLoadingReceipts = true
            errorMessage = null

            when (val r = repository.fetchReceipts(skip = 0, limit = pageSize)) {
                is ApiResult.Success -> {
                    receipts = r.data.receipts
                    receiptsTotal = r.data.total
                    receiptsHasMore = receipts.size < receiptsTotal
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            isLoadingReceipts = false
        }
    }

    fun loadMoreReceipts() {
        if (isLoadingMoreReceipts || !receiptsHasMore) return
        viewModelScope.launch {
            isLoadingMoreReceipts = true

            when (val r = repository.fetchReceipts(skip = receipts.size, limit = pageSize)) {
                is ApiResult.Success -> {
                    receipts = receipts + r.data.receipts
                    receiptsHasMore = receipts.size < receiptsTotal
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            isLoadingMoreReceipts = false
        }
    }

    fun refreshReceipts() {
        viewModelScope.launch {
            errorMessage = null

            when (val r = repository.fetchReceipts(skip = 0, limit = pageSize)) {
                is ApiResult.Success -> {
                    receipts = r.data.receipts
                    receiptsTotal = r.data.total
                    receiptsHasMore = receipts.size < receiptsTotal
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }
        }
    }

    fun deleteReceipt(id: String) {
        viewModelScope.launch {
            when (repository.deleteReceipt(id)) {
                is ApiResult.Success -> {
                    receipts = receipts.filter { it.id != id }
                    receiptsTotal -= 1
                }
                is ApiResult.Failure -> errorMessage = "Failed to delete receipt"
            }
        }
    }

    fun downloadReceipt(id: String) {
        viewModelScope.launch {
            when (repository.downloadReceipt(id)) {
                is ApiResult.Success -> successMessage = "Receipt download started"
                is ApiResult.Failure -> errorMessage = "Failed to download receipt"
            }
        }
    }

    fun parseReceipt(id: String) {
        viewModelScope.launch {
            isParsing = true

            when (repository.parseReceipt(id)) {
                is ApiResult.Success -> {
                    // Fetch updated receipt
                    when (val r = repository.fetchReceipt(id)) {
                        is ApiResult.Success -> {
                            parsedReceipt = r.data
                            receipts = receipts.map { if (it.id == id) r.data else it }
                        }
                        is ApiResult.Failure -> errorMessage = "Parse succeeded but failed to reload receipt"
                    }
                }
                is ApiResult.Failure -> errorMessage = "Failed to parse receipt"
            }

            isParsing = false
        }
    }

    fun linkReceiptToTransaction(receiptId: String, transactionId: String) {
        viewModelScope.launch {
            when (repository.linkReceipt(receiptId, transactionId)) {
                is ApiResult.Success -> {
                    when (val r = repository.fetchReceipt(receiptId)) {
                        is ApiResult.Success -> {
                            receipts = receipts.map { if (it.id == receiptId) r.data else it }
                        }
                        is ApiResult.Failure -> { /* Non-critical */ }
                    }
                    successMessage = "Receipt linked to transaction"
                }
                is ApiResult.Failure -> errorMessage = "Failed to link receipt"
            }
        }
    }

    // Reset upload state
    fun resetUploadState() {
        uploadedReceiptId = null
        parsedReceipt = null
        linkedTransactionId = ""
        isUploading = false
        isParsing = false
    }

    // Clear filters
    fun clearDateFilter() {
        dateFrom = null
        dateTo = null
        refreshRecords()
    }

    fun clearAmountFilter() {
        amountMin = ""
        amountMax = ""
        refreshRecords()
    }

    fun applyFilters() {
        refreshRecords()
    }

    fun dismissError() { errorMessage = null }
    fun dismissSuccess() { successMessage = null }

    // Helper
    private suspend fun fetchRecordsPage(skip: Int): ApiResult<RecordListResponse> {
        val dateFromStr = dateFrom?.let { dateFormatter.format(it) }
        val dateToStr = dateTo?.let { dateFormatter.format(it) }
        val minAmt = amountMin.toDoubleOrNull()
        val maxAmt = amountMax.toDoubleOrNull()

        return repository.fetchRecords(
            skip = skip,
            limit = pageSize,
            dateFrom = dateFromStr,
            dateTo = dateToStr,
            amountMin = minAmt,
            amountMax = maxAmt
        )
    }
}
